Tests/Common/Helpers/FakeURLProtocol.swift

```diff
@@ -0,0 +1,82 @@
+import Foundation
+
+/// A fake implementation of URLProtocol.
+/// This class is heavily adapted from https://jeroenscode.com/mocking-alamofire/.
+final class FakeURLProtocol: URLProtocol {
+  // Types of responses.
+  enum ResponseType {
+    // An error response and associated value.
+    case error(Error)
+
+    // A success response type, with the given response and ASCII encoded data.
+    case success(HTTPURLResponse, String)
+  }
+  static var responseType: ResponseType!
+
+  private lazy var session: URLSession = {
+    let configuration: URLSessionConfiguration = URLSessionConfiguration.ephemeral
+    return URLSession(configuration: configuration, delegate: self, delegateQueue: nil)
+  }()
+  private(set) var activeTask: URLSessionTask?
+
+  override class func canInit(with request: URLRequest) -> Bool {
+    return true
+  }
+
+  override class func canonicalRequest(for request: URLRequest) -> URLRequest {
+    return request
+  }
+
+  override class func requestIsCacheEquivalent(_ leftHandSide: URLRequest, to rightHandSide: URLRequest) -> Bool {
+    return false
+  }
+
+  override func startLoading() {
+    activeTask = session.dataTask(with: request.urlRequest!)
+    activeTask?.cancel()
+  }
+
+  override func stopLoading() {
+    activeTask?.cancel()
+  }
+}
+
+// MARK: - URLSessionDataDelegate
+
+extension FakeURLProtocol: URLSessionDataDelegate {
```

Comment:
What's the reasoning for using the class extensions?

Reply:
Generally when you conform to a protocol (`FakeURLProtocol` is conforming to `URLSessionDataDelegate`) you separate the functionality into an extension so it is clear what functions belong to what thing. 

Functionally, it is equivalent to just doing this at the class level.